src/templates/websocket.js

```diff
@@ -0,0 +1,87 @@
+(async (addr, hard) => {
+    addr = `ws://${addr}/live-server-ws`;
+    const sleep = (x) => new Promise(r => setTimeout(r, x));
+    const preload = async (url, requireSuccess) => {
+        const resp = await fetch(url, {cache: "reload"}); // reset cache
+        if (requireSuccess && (!resp.ok || resp.status !== 200)) { throw new Error(); }
+    }
+    /** Reset cache in link.href and strip scripts */
+    const preloadNode = (n, ps) => {
+        if (n.tagName === "SCRIPT" && n.src) { ps.push(preload(n.src, false)); return; }
+        if (n.tagName === "LINK" && n.href) { ps.push(preload(n.href, false)); return; }
+        let c = n.firstChild; while (c) { let nc = c.nextSibling; preloadNode(c, ps); c = nc; }
+    };
+    let reloading = false;
+    let scheduled = false;
+    async function reload() {
+        if(reloading) {
+            scheduled = true;
+            return;
+        }
+        let ifr;
+        let interval = 0;
+        reloading = true;
+        while(true) {
+            scheduled = false;
+            try {
+                const url = location.origin + location.pathname;
+                const ps = [];
+                preloadNode(document.head, ps);
+                preloadNode(document.body, ps);
+                await Promise.allSettled(ps);
+                await new Promise((resolve) => {
+                    ifr = document.createElement("iframe");
+                    ifr.src = url + "?reload";
+                    ifr.style.display = "none";
+                    ifr.onload = resolve;
+                    document.body.appendChild(ifr);
+                });
+                const meta = ifr.contentDocument.head.lastChild;
+                if (meta.tagName !== "META" || meta.name !== "live-server" || meta.content !== "reload") {
+                    throw new Error();
+                }
+                if (hard) {
+                    location.reload();
+                }
+                document.head.replaceWith(ifr.contentDocument.head);
+                document.body.replaceWith(ifr.contentDocument.body);
+                ifr.remove();
+                if (!scheduled) { 
```

Comment:
When another reload message is received while we were preloading the page. In this case, we will abandon this reload and retry in the next iteration of the loop

Reply:
In line 32 we set `scheduled = true` indeed, but then without awaiting any time-consuming operation we set `scheduled = false` back in line 38, and started awaiting some promises. So when in the second `reload()` it's awaiting the promises and in the first `reload()` it just finished loading the iframe and is checking the `scheduled`, the result will be false and it will still continue `location.reload()`, right?

Reply:
on the first `reload()` call, `reloading` is false and set to true and the function enters the loop synchronously, where it starts with scheduled = false and waits on the preload promises. If during the preload, another `reload()` is called, it will see `reloading` is true and set `scheduled` to true and return. When the preload finishes, it will see another `reload()` has been called during this time because scheduled is true, so it abandons the current attempt and starts over to make sure it gets the latest update